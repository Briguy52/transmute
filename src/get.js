import curry from "./curry";
import { Iterable } from "immutable";

/**
 * Retrieve the value at `key` from `subject`.
 *
 * @example
 * // returns 1
 * get('one', Map({one: 1, two: 2, three: 3}))
 *
 * @param  {any} key to lookup in `subject`.
 * @param  {Iterable|Object} subject in which to look up `key`.
 * @return {any} the value at `key`.
 */
function get(key, subject) {
  if (subject === null || subject === undefined) {
    return undefined;
  }
  if (Iterable.isIterable(subject)) {
    return subject.get(key);
  }
  return subject[key];
}

export default curry(get);
