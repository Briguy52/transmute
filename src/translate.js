import curry from "./curry";
import { Map } from "immutable";
import get from "./get";
import getIn from "./getIn";
import isRecord from "./isRecord";
import reduce from "./reduce";
import set from "./set";

const getOp = get.operation;
const getInOp = getIn.operation;
const reduceOp = reduce.operation;
const setOp = set.operation;

function runTransform(transform, newKey, subject) {
  if (typeof transform === "function") {
    return transform(subject, newKey);
  }
  if (transform === true) {
    return getOp(newKey, subject);
  }
  if (Array.isArray(transform)) {
    return getInOp(transform, subject);
  }
  return getOp(transform, subject);
}

function translate(translation, subject) {
  const result = isRecord(subject) ? Map() : new subject.constructor();
  return reduceOp(
    result,
    (acc, transform, newKey) =>
      setOp(runTransform(transform, newKey, subject), newKey, acc),
    translation
  );
}

export default curry(translate);
