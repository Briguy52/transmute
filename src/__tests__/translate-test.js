import { Map, OrderedMap } from "immutable";
import get from "../get";
import translate from "../translate";

describe("transmute/translate", () => {
  it("takes keys flagged `true`", () => {
    const translation = translate({ one: true });
    expect(translation(Map({ one: 1, two: 2, three: 3 }))).toMatchSnapshot();
  });

  it("renames keys with string values", () => {
    const translation = translate({ first: "one" });
    expect(translation(Map({ one: 1, two: 2, three: 3 }))).toMatchSnapshot();
  });

  it("applies functions to the key", () => {
    const translation = translate({
      second: s => s.get("one") + 1
    });
    expect(translation(Map({ one: 1, two: 2, three: 3 }))).toMatchSnapshot();
  });

  describe("compatibility", () => {
    const getThree = get("three");
    const translation = translate({
      one: true,
      second: "two",
      three: s => getThree(s) + 1
    });

    it("works with Maps", () => {
      expect(translation(Map({ one: 1, two: 2, three: 3 }))).toMatchSnapshot();
    });

    it("works with Objects", () => {
      expect(translation({ one: 1, two: 2, three: 3 })).toMatchSnapshot();
    });

    it("works with OrderedMaps", () => {
      expect(
        translation(OrderedMap({ one: 1, two: 2, three: 3 }))
      ).toMatchSnapshot();
    });
  });
});
