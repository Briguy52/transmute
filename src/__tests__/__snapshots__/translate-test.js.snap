// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transmute/translate applies functions to the key 1`] = `
Immutable.Map {
  second: 2,
}
`;

exports[`transmute/translate compatibility works with Maps 1`] = `
Immutable.Map {
  one: 1,
  second: 2,
  three: 4,
}
`;

exports[`transmute/translate compatibility works with Objects 1`] = `
Object {
  "one": 1,
  "second": 2,
  "three": 4,
}
`;

exports[`transmute/translate compatibility works with OrderedMaps 1`] = `
Immutable.OrderedMap {
  one: 1,
  second: 2,
  three: 4,
}
`;

exports[`transmute/translate renames keys with string values 1`] = `
Immutable.Map {
  first: 1,
}
`;

exports[`transmute/translate takes keys flagged \`true\` 1`] = `
Immutable.Map {
  one: 1,
}
`;
